ocaml/xapi/xapi_pusb_helpers.ml

```diff
@@ -118,3 +118,16 @@ let find_or_create ~__context pusb =
   let name_label = "Group of " ^ pusb_rec.Db_actions.pUSB_vendor_id ^ " " ^ pusb_rec.Db_actions.pUSB_product_id ^ " USBs" in
   let group = Xapi_usb_group.create ~__context ~name_label ~name_description:"" ~other_config:[] in
   group
+
+let destroy_pusb ~__context pusb =
+  let usb_group = Db.PUSB.get_USB_group ~__context ~self:pusb in
+  let vusbs = Db.USB_group.get_VUSBs ~__context ~self:usb_group in
+  List.iter (fun vusb ->
+    let currently_attached = Db.VUSB.get_currently_attached ~__context ~self:vusb in
+    if currently_attached then
+      Helpers.call_api_functions ~__context (fun rpc session_id ->
+        Client.Client.VUSB.unplug rpc session_id vusb)
+  ) vusbs;
+  List.iter (fun vusb -> Db.VUSB.destroy ~__context ~self:vusb) vusbs;
+  Db.PUSB.destroy ~__context ~self:pusb;
+  Db.USB_group.destroy ~__context ~self:usb_group
```

Comment:
I assume the old version was catching the errors thus running the `PUSB.destroy` and `USB_group.destroy` functions in any case, while the current will do it only if the VUSB unplug&destroy are successful. Is that the change that happened? In this case, could you add a comment explaining that this function has to fail when an unplug fails, to prevent an incorrect PUSB and USB_group destruction?

Is there a reason we iterate twice on the vusbs instead of using a unique block?
```ocaml
List.iter (fun vusb ->
    let currently_attached = Db.VUSB.get_currently_attached ~__context ~self:vusb in
    if currently_attached then
      Helpers.call_api_functions ~__context (fun rpc session_id ->
        Client.Client.VUSB.unplug rpc session_id vusb);

    Db.VUSB.destroy ~__context ~self:vusb
) vusbs;

Reply:
yes @mseri  Only when unplug vusb successfully , then we will destroy PUSB and USB_group. If we catch a exception when unplug vusb. We should skip destroy PUSB and USB_group and continue other PUSB scan

Reply:
As @mseri says, let's merge the `List.iter`s.

Reply:
@robhoes , just updated and add some commit comments in the second commit.